File: lib/compiler/lexer.js
/**
 * Lexer (Tokenizer)
 * Converts source code into a stream of tokens
 */

export const TokenType = {
  // Keywords
  LET: 'LET',
  IF: 'IF',
  ELSE: 'ELSE',
  WHILE: 'WHILE',
  PRINT: 'PRINT',

  // Identifiers and literals
  IDENTIFIER: 'IDENTIFIER',
  NUMBER: 'NUMBER',
  STRING: 'STRING',

  // Operators
  PLUS: 'PLUS',
  MINUS: 'MINUS',
  MULTIPLY: 'MULTIPLY',
  DIVIDE: 'DIVIDE',
  MODULO: 'MODULO',
  ASSIGN: 'ASSIGN',

  // Comparison operators
  EQUAL: 'EQUAL',
  NOT_EQUAL: 'NOT_EQUAL',
  LESS_THAN: 'LESS_THAN',
  GREATER_THAN: 'GREATER_THAN',
  LESS_EQUAL: 'LESS_EQUAL',
  GREATER_EQUAL: 'GREATER_EQUAL',

  // Delimiters
  SEMICOLON: 'SEMICOLON',
  LPAREN: 'LPAREN',
  RPAREN: 'RPAREN',
  LBRACE: 'LBRACE',
  RBRACE: 'RBRACE',

  // Special
  EOF: 'EOF',
  NEWLINE: 'NEWLINE',
};

const KEYWORDS = {
  'let': TokenType.LET,
  'if': TokenType.IF,
  'else': TokenType.ELSE,
  'while': TokenType.WHILE,
  'print': TokenType.PRINT,
};

export class Token {
  constructor(type, value, line, column) {
    this.type = type;
    this.value = value;
    this.line = line;
    this.column = column;
  }

  toString() {
    return `Token(${this.type}, ${JSON.stringify(this.value)}, ${this.line}:${this.column})`;
  }
}

export class Lexer {
  constructor(source) {
    this.source = source;
    this.position = 0;
    this.line = 1;
    this.column = 1;
    this.tokens = [];
  }

  currentChar() {
    return this.position < this.source.length ? this.source[this.position] : null;
  }

  peek(offset = 1) {
    const pos = this.position + offset;
    return pos < this.source.length ? this.source[pos] : null;
  }

  advance() {
    if (this.currentChar() === '\n') {
      this.line++;
      this.column = 1;
    } else {
      this.column++;
    }
    this.position++;
  }

  skipWhitespace() {
    while (this.currentChar() && /\s/.test(this.currentChar()) && this.currentChar() !== '\n') {
      this.advance();
    }
  }

  skipComment() {
    if (this.currentChar() === '/' && this.peek() === '/') {
      while (this.currentChar() && this.currentChar() !== '\n') {
        this.advance();
      }
    }
  }

  readNumber() {
    const startLine = this.line;
    const startColumn = this.column;
    let numStr = '';
    let hasDecimal = false;

    while (this.currentChar() && (/\d/.test(this.currentChar()) || this.currentChar() === '.')) {
      if (this.currentChar() === '.') {
        if (hasDecimal) break;
        hasDecimal = true;
      }
      numStr += this.currentChar();
      this.advance();
    }

    return new Token(TokenType.NUMBER, parseFloat(numStr), startLine, startColumn);
  }

  readString() {
    const startLine = this.line;
    const startColumn = this.column;
    const quote = this.currentChar();
    this.advance(); // skip opening quote

    let str = '';
    while (this.currentChar() && this.currentChar() !== quote) {
      if (this.currentChar() === '\\') {
        this.advance();
        const escapeChar = this.currentChar();
        switch (escapeChar) {
          case 'n': str += '\n'; break;
          case 't': str += '\t'; break;
          case '\\': str += '\\'; break;
          case quote: str += quote; break;
          default: str += escapeChar;
        }
        this.advance();
      } else {
        str += this.currentChar();
        this.advance();
      }
    }

    if (this.currentChar() === quote) {
      this.advance(); // skip closing quote
    } else {
      throw new Error(`Unterminated string at ${startLine}:${startColumn}`);
    }

    return new Token(TokenType.STRING, str, startLine, startColumn);
  }

  readIdentifier() {
    const startLine = this.line;
    const startColumn = this.column;
    let identifier = '';

    while (this.currentChar() && /[a-zA-Z0-9_]/.test(this.currentChar())) {
      identifier += this.currentChar();
      this.advance();
    }

    const type = KEYWORDS[identifier] || TokenType.IDENTIFIER;
    return new Token(type, identifier, startLine, startColumn);
  }

  tokenize() {
    this.tokens = [];

    while (this.currentChar()) {
      const startLine = this.line;
      const startColumn = this.column;

      // Skip whitespace
      if (/\s/.test(this.currentChar()) && this.currentChar() !== '\n') {
        this.skipWhitespace();
        continue;
      }

      // Skip comments
      if (this.currentChar() === '/' && this.peek() === '/') {
        this.skipComment();
        continue;
      }

      // Newline
      if (this.currentChar() === '\n') {
        this.tokens.push(new Token(TokenType.NEWLINE, '\n', startLine, startColumn));
        this.advance();
        continue;
      }

      // Numbers
      if (/\d/.test(this.currentChar())) {
        this.tokens.push(this.readNumber());
        continue;
      }

      // Strings
      if (this.currentChar() === '"' || this.currentChar() === "'") {
        this.tokens.push(this.readString());
        continue;
      }

      // Identifiers and keywords
      if (/[a-zA-Z_]/.test(this.currentChar())) {
        this.tokens.push(this.readIdentifier());
        continue;
      }

      // Two-character operators
      if (this.currentChar() === '=' && this.peek() === '=') {
        this.tokens.push(new Token(TokenType.EQUAL, '==', startLine, startColumn));
        this.advance();
        this.advance();
        continue;
      }

      if (this.currentChar() === '!' && this.peek() === '=') {
        this.tokens.push(new Token(TokenType.NOT_EQUAL, '!=', startLine, startColumn));
        this.advance();
        this.advance();
        continue;
      }

      if (this.currentChar() === '<' && this.peek() === '=') {
        this.tokens.push(new Token(TokenType.LESS_EQUAL, '<=', startLine, startColumn));
        this.advance();
        this.advance();
        continue;
      }

      if (this.currentChar() === '>' && this.peek() === '=') {
        this.tokens.push(new Token(TokenType.GREATER_EQUAL, '>=', startLine, startColumn));
        this.advance();
        this.advance();
        continue;
      }

      // Single-character tokens
      const char = this.currentChar();
      let token = null;

      switch (char) {
        case '+': token = new Token(TokenType.PLUS, '+', startLine, startColumn); break;
        case '-': token = new Token(TokenType.MINUS, '-', startLine, startColumn); break;
        case '*': token = new Token(TokenType.MULTIPLY, '*', startLine, startColumn); break;
        case '/': token = new Token(TokenType.DIVIDE, '/', startLine, startColumn); break;
        case '%': token = new Token(TokenType.MODULO, '%', startLine, startColumn); break;
        case '=': token = new Token(TokenType.ASSIGN, '=', startLine, startColumn); break;
        case '<': token = new Token(TokenType.LESS_THAN, '<', startLine, startColumn); break;
        case '>': token = new Token(TokenType.GREATER_THAN, '>', startLine, startColumn); break;
        case ';': token = new Token(TokenType.SEMICOLON, ';', startLine, startColumn); break;
        case '(': token = new Token(TokenType.LPAREN, '(', startLine, startColumn); break;
        case ')': token = new Token(TokenType.RPAREN, ')', startLine, startColumn); break;
        case '{': token = new Token(TokenType.LBRACE, '{', startLine, startColumn); break;
        case '}': token = new Token(TokenType.RBRACE, '}', startLine, startColumn); break;
        default:
          throw new Error(`Unexpected character '${char}' at ${startLine}:${startColumn}`);
      }

      if (token) {
        this.tokens.push(token);
        this.advance();
      }
    }

    this.tokens.push(new Token(TokenType.EOF, null, this.line, this.column));
    return this.tokens;
  }
}

File: backend/compiler/lexer.py
from dataclasses import dataclass
from typing import List

TOKEN_TYPES = (
    'NUMBER',
    'PLUS',
    'MINUS',
    'TIMES',
    'DIVIDE',
    'MODULO',
    'LPAREN',
    'RPAREN',
    'LBRACE',
    'RBRACE',
    'SEMICOLON',
    'ASSIGN',
    'EQ',
    'NE',
    'LT',
    'GT',
    'LE',
    'GE',
    'ID',
    'LET',
    'IF',
    'ELSE',
    'WHILE',
    'PRINT',
)

RESERVED_KEYWORDS = {
    'let': 'LET',
    'if': 'IF',
    'else': 'ELSE',
    'while': 'WHILE',
    'print': 'PRINT',
}


@dataclass(frozen=True)
class Token:
    token_type: str
    value: object
    line: int
    position: int


SINGLE_CHAR_TOKEN_TYPES = {
    '+': 'PLUS',
    '-': 'MINUS',
    '*': 'TIMES',
    '/': 'DIVIDE',
    '%': 'MODULO',
    '(': 'LPAREN',
    ')': 'RPAREN',
    '{': 'LBRACE',
    '}': 'RBRACE',
    ';': 'SEMICOLON',
    '=': 'ASSIGN',
    '<': 'LT',
    '>': 'GT',
}

DOUBLE_CHAR_TOKEN_TYPES = {
    '==': 'EQ',
    '!=': 'NE',
    '<=': 'LE',
    '>=': 'GE',
}


def tokenize_objects(code: str) -> List[Token]:
    current_index = 0
    current_line = 1
    code_length = len(code)
    token_list: List[Token] = []

    while current_index < code_length:
        current_char = code[current_index]

        if current_char in ' \t\r':
            current_index += 1
            continue

        if current_char == '\n':
            current_line += 1
            current_index += 1
            continue

        if current_char == '/' and current_index + 1 < code_length and code[current_index + 1] == '/':
            current_index += 2
            while current_index < code_length and code[current_index] != '\n':
                current_index += 1
            continue

        if current_index + 1 < code_length:
            two_char_token_text = code[current_index:current_index + 2]
            token_type = DOUBLE_CHAR_TOKEN_TYPES.get(two_char_token_text)
            if token_type:
                token_list.append(Token(token_type=token_type, value=two_char_token_text, line=current_line, position=current_index))
                current_index += 2
                continue

        if current_char.isdigit():
            number_start_index = current_index
            while current_index < code_length and code[current_index].isdigit():
                current_index += 1
            number_text = code[number_start_index:current_index]
            token_list.append(Token(token_type='NUMBER', value=int(number_text), line=current_line, position=number_start_index))
            continue

        if current_char.isalpha() or current_char == '_':
            identifier_start_index = current_index
            while current_index < code_length and (code[current_index].isalnum() or code[current_index] == '_'):
                current_index += 1
            identifier_text = code[identifier_start_index:current_index]
            token_type = RESERVED_KEYWORDS.get(identifier_text, 'ID')
            token_list.append(Token(token_type=token_type, value=identifier_text, line=current_line, position=identifier_start_index))
            continue

        single_char_token_type = SINGLE_CHAR_TOKEN_TYPES.get(current_char)
        if single_char_token_type:
            token_list.append(Token(token_type=single_char_token_type, value=current_char, line=current_line, position=current_index))
            current_index += 1
            continue

        raise SyntaxError(f"Illegal character '{current_char}' at line {current_line}")

    return token_list


def tokenize(code: str):
    return [
        {
            'type': token.token_type,
            'value': str(token.value),
            'line': token.line,
            'position': token.position,
        }
        for token in tokenize_objects(code)
    ]


tokens = TOKEN_TYPES
reserved = RESERVED_KEYWORDS
